File: spring-backend/src/main/java/com/simmortal/memorial/MemorialController.java
package com.simmortal.memorial;

import com.simmortal.ai.AiService;
import com.simmortal.util.AssetService;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RequestPart;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

@RestController
@RequestMapping("memorial")
public class MemorialController {
  private final MemorialService memorialService;
  private final AiService aiService;
  private final AssetService assetService;

  public MemorialController(MemorialService memorialService, AiService aiService, AssetService assetService) {
    this.memorialService = memorialService;
    this.aiService = aiService;
    this.assetService = assetService;
  }

  @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
  public Object createMemorial(
      @RequestHeader("X-User-Id") String userId,
      @RequestPart("data") Map<String, Object> request,
      @RequestPart("image") MultipartFile image
  ) {
    return memorialService.createMemorial(userId, request, image);
  }

  @GetMapping("search")
  public Object searchMemorials(
      @RequestParam("query") String query,
      @RequestParam("limit") Integer limit
  ) {
    return memorialService.searchMemorialByName(query, limit);
  }

  @GetMapping("featured")
  public Object getFeaturedMemorials(@RequestParam("limit") Integer limit) {
    return memorialService.getFeaturedMemorials(limit);
  }

  @PostMapping("preview/{memorialId}")
  public Object createMemorialPreview(
      @PathVariable String memorialId,
      @RequestBody Map<String, Object> overrides
  ) {
    return memorialService.createMemorialPreview(memorialId, overrides);
  }

  @GetMapping("publish-preview/{memorialId}")
  public Object getMemorialPreview(@PathVariable String memorialId) {
    return memorialService.getPublishMemorialPreview(memorialId);
  }

  @PostMapping(value = "transcribe", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
  public Map<String, Object> transcribeAudio(@RequestPart("audio") MultipartFile audio) {
    String transcription = aiService.transcribeAudio(audio);
    Map<String, Object> response = new HashMap<>();
    response.put("transcription", transcription);
    return response;
  }

  @PostMapping("check-slug")
  public Map<String, Object> checkSlugAvailability(@RequestBody CheckSlugAvailabilityRequest request) {
    boolean isAvailable = memorialService.checkSlugAvailability(request.slug(), request.memorialId());
    Map<String, Object> response = new HashMap<>();
    response.put("isAvailable", isAvailable);
    return response;
  }

  @GetMapping("premium-prices")
  public Object getPremiumSubscriptionPrices() {
    return memorialService.getPremiumSubscriptionPrices();
  }

  @PostMapping("purchase-premium/{memorialId}")
  public Object purchasePremiumMemorial(
      @RequestHeader("X-User-Id") String userId,
      @PathVariable String memorialId,
      @RequestBody PurchasePremiumMemorialRequest request
  ) {
    return memorialService.createPremiumSubscriptionLink(userId, memorialId, request.period());
  }

  @PostMapping("validate-subscription/{sessionId}")
  public Object validateSubscriptionResult(
      @PathVariable String sessionId,
      @RequestBody ValidateSubscriptionRequest request
  ) {
    return memorialService.validateSubscriptionResult(sessionId, request.publishOnValidation());
  }

  @PostMapping("publish-free/{memorialId}")
  public Object publishFreeMemorial(
      @RequestHeader("X-User-Id") String userId,
      @PathVariable String memorialId
  ) {
    return memorialService.publishFreeMemorial(userId, memorialId);
  }

  @GetMapping("published/{slug}")
  public Object getPublishedMemorial(
      @PathVariable String slug,
      @RequestHeader(value = "X-User-Id", required = false) String userId
  ) {
    return memorialService.getPublishedMemorial(slug, userId);
  }

  @GetMapping("published/id/{id}")
  public Object getPublishedMemorialById(@PathVariable String id) {
    return memorialService.getPublishedMemorialById(id);
  }

  @PostMapping("published/{slug}/view")
  public void logPublishedMemorialView(@PathVariable String slug, HttpServletRequest request) {
    memorialService.incrementMemorialViewBySlug(slug, request);
  }

  @PostMapping("published/{slug}/like")
  public void likePublishedMemorial(
      @PathVariable String slug,
      @RequestHeader("X-User-Id") String userId
  ) {
    String memorialId = memorialService.getMemorialIdBySlug(slug);
    if (memorialId == null) {
      throw new IllegalStateException("Memorial not found");
    }
    memorialService.likeMemorial(memorialId, userId);
  }

  @DeleteMapping("published/{slug}/like")
  public void unlikePublishedMemorial(
      @PathVariable String slug,
      @RequestHeader("X-User-Id") String userId
  ) {
    String memorialId = memorialService.getMemorialIdBySlug(slug);
    if (memorialId == null) {
      throw new IllegalStateException("Memorial not found");
    }
    memorialService.unlikeMemorial(memorialId, userId);
  }

  @GetMapping("owned/{memorialId}")
  public Object getMemorial(
      @RequestHeader("X-User-Id") String userId,
      @PathVariable String memorialId,
      @RequestParam(value = "recommendSlug", required = false) Boolean recommendSlug
  ) {
    return memorialService.getMemorial(userId, memorialId, recommendSlug);
  }

  @GetMapping("owned-preview")
  public Object getOwnedMemorials(@RequestHeader("X-User-Id") String userId) {
    return memorialService.getOwnedMemorialPreviews(userId);
  }

  @PatchMapping(value = "owned/{memorialId}", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
  public Object updateMemorial(
      @RequestHeader("X-User-Id") String userId,
      @PathVariable String memorialId,
      @RequestPart("data") Map<String, Object> request,
      @RequestPart(value = "image", required = false) MultipartFile image,
      @RequestPart(value = "coverImage", required = false) MultipartFile coverImage
  ) {
    return memorialService.updateMemorial(userId, memorialId, request, image, coverImage);
  }

  @DeleteMapping("owned/{memorialId}")
  public Object deleteMemorial(
      @RequestHeader("X-User-Id") String userId,
      @PathVariable String memorialId
  ) {
    return memorialService.deleteMemorial(userId, memorialId);
  }

  @PostMapping("published/flag")
  public void flagPublishedMemorial(
      @RequestHeader("X-User-Id") String userId,
      @RequestBody CreateMemorialFlagRequest request
  ) {
    if (request.type() != MemorialFlagType.CONDOLENCE_REPORT
        && request.type() != MemorialFlagType.MEMORY_REPORT
        && request.type() != MemorialFlagType.MEMORIAL_REPORT
        && request.type() != MemorialFlagType.DONATION_REPORT) {
      throw new IllegalArgumentException("Invalid flag type");
    }
    memorialService.createMemorialFlag(userId, request.type(), request.referenceId(), request.reason());
  }

  @GetMapping("published/{slug}/flag")
  public Object getMemorialFlags(
      @RequestHeader("X-User-Id") String userId,
      @PathVariable String slug
  ) {
    return memorialService.getMemorialFlags(userId, slug);
  }

  @PatchMapping("flag/{flagId}")
  public void handleMemorialFlag(
      @RequestHeader("X-User-Id") String userId,
      @PathVariable String flagId,
      @RequestBody Map<String, String> request
  ) {
    MemorialFlagStatus status = MemorialFlagStatus.fromValue(request.get("status"));
    memorialService.handleMemorialUpdateFlag(userId, flagId, status);
  }

  @GetMapping("ai-greeting/{memorialId}")
  public Object getMemorialAiGreeting(
      @PathVariable String memorialId,
      @RequestHeader("X-User-Id") String userId
  ) {
    memorialService.verifyMemorialOwnership(userId, memorialId);
    return aiService.getAiGreeting(memorialId);
  }

  @DeleteMapping("ai-greeting/{memorialId}")
  public void deleteMemorialAiVoice(
      @PathVariable String memorialId,
      @RequestHeader("X-User-Id") String userId
  ) {
    memorialService.verifyMemorialOwnership(userId, memorialId);
    aiService.resetAiGreetingCreation(memorialId);
  }

  @PostMapping(value = "ai-greeting/voice/{memorialId}", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
  public Map<String, Object> createMemorialVoice(
      @PathVariable String memorialId,
      @RequestHeader("X-User-Id") String userId,
      @RequestPart("files") List<MultipartFile> files,
      @RequestParam(value = "locale", required = false) String locale
  ) {
    memorialService.verifyMemorialOwnership(userId, memorialId);
    String voiceId = aiService.createMemorialVoice(memorialId, files);
    String localeToUse = ("tr".equals(locale) || "en".equals(locale)) ? locale : "en";
    String path = aiService.createMemorialGreetingAudio(memorialId, voiceId, localeToUse);
    Map<String, Object> response = new HashMap<>();
    response.put("path", assetService.generateAssetUrl(path));
    return response;
  }

  @PostMapping(value = "ai-greeting/image/{memorialId}", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
  public Map<String, Object> uploadAiGreetingImage(
      @PathVariable String memorialId,
      @RequestHeader("X-User-Id") String userId,
      @RequestPart("image") MultipartFile image
  ) {
    memorialService.verifyMemorialOwnership(userId, memorialId);
    String path = aiService.uploadAiGreetingImage(memorialId, image);
    Map<String, Object> response = new HashMap<>();
    response.put("path", assetService.generateAssetUrl(path));
    return response;
  }

  @GetMapping("published/{slug}/top-contributors/candles")
  public Object getTopCandleContributors(@PathVariable String slug) {
    return memorialService.getTopCandleContributors(slug);
  }

  @GetMapping("published/{slug}/top-contributors/donations")
  public Object getTopDonors(@PathVariable String slug) {
    return memorialService.getTopDonors(slug);
  }

  @GetMapping("published/{slug}/top-contributors/flowers")
  public Object getTopFlowerContributors(@PathVariable String slug) {
    return memorialService.getTopFlowerContributors(slug);
  }

  @GetMapping("published/{slug}/top-contributors/trees")
  public Object getTopTreePlanters(@PathVariable String slug) {
    return memorialService.getTopTreePlanters(slug);
  }
}

File: Backend-main/src/memorial/memorial.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Patch,
  Post,
  Query,
  Req,
  UploadedFile,
  UploadedFiles,
  UseInterceptors,
} from '@nestjs/common';
import { FileFieldsInterceptor, FileInterceptor, FilesInterceptor } from '@nestjs/platform-express';
import { PublicAccess, Session } from 'supertokens-nestjs';
import { SessionContainer } from 'supertokens-node/recipe/session';
import { Express, Request } from 'express';
import { MemorialService } from './memorial.service';
import {
  Memorial,
  MemorialFlag,
  MemorialFlagStatus,
  MemorialFlagType,
  OwnedMemorialPreview,
  PublishedMemorial,
  PublishMemorialPreview,
  TopContributor,
} from './interface/memorial.interface';
import { AIService } from 'src/ai/ai.service';
import { MemoryService } from './memory/memory.service';
import { PublicWithOptionalSession } from 'src/guard/optional-session.guard';
import { CacheInterceptor, CacheTTL } from '@nestjs/cache-manager';
import {
  CheckSlugAvailabilityRequest,
  CreateMemorialFlagRequest,
  CreateMemorialRequest,
  PremiumPricesResponse,
  PurchasePremiumMemorialRequest,
  ValidatePurchaseResultResponse,
} from './interface/memorial.dto';
import { generateAssetUrl } from 'src/util/asset';

@Controller('memorial')
export class MemorialController {
  constructor(
    private readonly memorialService: MemorialService,
    private readonly memoryService: MemoryService,
    private readonly aiService: AIService,
  ) {}

  @Post()
  @UseInterceptors(
    FileInterceptor('image', {
      limits: {
        files: 1,
      },
    }),
  )
  async createMemorial(
    @Session() session: SessionContainer,
    @Body() request: CreateMemorialRequest,
    @UploadedFile() image: Express.Multer.File,
  ) {
    const userId = session.getUserId();
    return await this.memorialService.createMemorial(userId, request, image);
  }

  @PublicAccess()
  @Get('search')
  async searchMemorials(
    @Query('query') query: string,
    @Query('limit') limit: number,
  ): Promise<PublishedMemorial[]> {
    return this.memorialService.searchMemorialByName(query, limit);
  }

  @PublicAccess()
  @Get('featured')
  async getFeaturedMemorials(@Query('limit') limit: number): Promise<PublishedMemorial[]> {
    return this.memorialService.getFeaturedMemorials(limit);
  }

  @Post('preview/:memorialId')
  async createMemorialPreview(
    @Param('memorialId') memorialId: string,
    @Body() overrides: Partial<Memorial>,
  ) {
    return await this.memorialService.createMemorialPreview(memorialId, overrides);
  }

  @Get('publish-preview/:memorialId')
  async getMemorialPreview(
    @Param('memorialId') memorialId: string,
  ): Promise<PublishMemorialPreview> {
    return await this.memorialService.getPublishMemorialPreview(memorialId);
  }

  @Post('transcribe')
  @UseInterceptors(
    FileInterceptor('audio', {
      limits: {
        files: 1,
      },
    }),
  )
  async transcribeAudio(@UploadedFile() audio: Express.Multer.File) {
    const transcription = await this.aiService.transcribeAudio(audio);
    return { transcription };
  }

  @Post('check-slug')
  async checkSlugAvailability(@Body() request: CheckSlugAvailabilityRequest) {
    const isAvailable = await this.memorialService.checkSlugAvailability(
      request.slug,
      request.memorialId,
    );
    return { isAvailable };
  }

  @PublicAccess()
  @Get('premium-prices')
  @UseInterceptors(CacheInterceptor)
  @CacheTTL(300000) // Cache for 5 minutes
  async getPremiumSubscriptionPrices(): Promise<PremiumPricesResponse> {
    return this.memorialService.getPremiumSubscriptionPrices();
  }

  @Post('purchase-premium/:memorialId')
  async purchasePremiumMemorial(
    @Session() session: SessionContainer,
    @Param('memorialId') memorialId: string,
    @Body() { period }: PurchasePremiumMemorialRequest,
  ) {
    const userId = session.getUserId();
    return await this.memorialService.createPremiumSubscriptionLink(userId, memorialId, period);
  }

  @Post('validate-subscription/:sessionId')
  async validateSubscriptionResult(
    @Param('sessionId') sessionId: string,
    @Body() { publishOnValidation }: { publishOnValidation?: boolean },
  ): Promise<ValidatePurchaseResultResponse> {
    return this.memorialService.validateSubscriptionResult(sessionId, publishOnValidation);
  }

  @Post('publish-free/:memorialId')
  async publishFreeMemorial(
    @Session() session: SessionContainer,
    @Param('memorialId') memorialId: string,
  ) {
    const userId = session.getUserId();
    return await this.memorialService.publishFreeMemorial(userId, memorialId);
  }

  @PublicWithOptionalSession()
  @Get('published/:slug')
  async getPublishedMemorial(@Param('slug') slug: string, @Session() session?: SessionContainer) {
    const userId = session?.getUserId();
    return await this.memorialService.getPublishedMemorial(slug, userId);
  }

  @PublicAccess()
  @Get('published/id/:id')
  async getPublishedMemorialById(@Param('id') id: string): Promise<PublishedMemorial> {
    return await this.memorialService.getPublishedMemorialById(id);
  }

  @PublicAccess()
  @Post('published/:slug/view')
  async logPublishedMemorialView(@Param('slug') slug: string, @Req() req: Request): Promise<void> {
    return this.memorialService.incrementMemorialViewBySlug(slug, req);
  }

  @Post('published/:slug/like')
  async likePublishedMemorial(
    @Param('slug') slug: string,
    @Session() session: SessionContainer,
  ): Promise<void> {
    const userId = session.getUserId();
    const memorialId = await this.memorialService.getMemorialIdBySlug(slug);
    if (!memorialId) {
      throw new Error('Memorial not found');
    }
    return this.memorialService.likeMemorial(memorialId, userId);
  }

  @Delete('published/:slug/like')
  async unlikePublishedMemorial(
    @Param('slug') slug: string,
    @Session() session: SessionContainer,
  ): Promise<void> {
    const userId = session.getUserId();
    const memorialId = await this.memorialService.getMemorialIdBySlug(slug);
    if (!memorialId) {
      throw new Error('Memorial not found');
    }
    return this.memorialService.unlikeMemorial(memorialId, userId);
  }

  @Get('owned/:memorialId')
  async getMemorial(
    @Session() session: SessionContainer,
    @Param('memorialId') memorialId: string,
    @Query('recommendSlug') recommendSlug?: boolean,
  ): Promise<Memorial> {
    const userId = session.getUserId();
    return await this.memorialService.getMemorial(userId, memorialId, recommendSlug);
  }

  @Get('owned-preview')
  async getOwnedMemorials(@Session() session: SessionContainer): Promise<OwnedMemorialPreview[]> {
    const userId = session.getUserId();
    return await this.memorialService.getOwnedMemorialPreviews(userId);
  }

  @Patch('owned/:memorialId')
  @UseInterceptors(
    FileFieldsInterceptor([
      { name: 'image', maxCount: 1 },
      { name: 'coverImage', maxCount: 1 },
    ]),
  )
  async updateMemorial(
    @Session() session: SessionContainer,
    @Param('memorialId') memorialId: string,
    @Body() request: Partial<Memorial>,
    @UploadedFiles() files?: { image?: Express.Multer.File[]; coverImage?: Express.Multer.File[] },
  ) {
    const userId = session.getUserId();
    return await this.memorialService.updateMemorial(userId, memorialId, request, {
      image: files?.image?.[0],
      coverImage: files?.coverImage?.[0],
    });
  }

  @Delete('owned/:memorialId')
  async deleteMemorial(
    @Session() session: SessionContainer,
    @Param('memorialId') memorialId: string,
  ) {
    const userId = session.getUserId();
    return await this.memorialService.deleteMemorial(userId, memorialId);
  }

  @Post('published/flag')
  async flagPublishedMemorial(
    @Session() session: SessionContainer,
    @Body() request: CreateMemorialFlagRequest,
  ): Promise<void> {
    const userId = session.getUserId();
    if (
      request.type !== MemorialFlagType.CONDOLENCE_REPORT &&
      request.type !== MemorialFlagType.MEMORY_REPORT &&
      request.type !== MemorialFlagType.MEMORIAL_REPORT &&
      request.type !== MemorialFlagType.DONATION_REPORT
    ) {
      throw new HttpException('Invalid flag type', HttpStatus.BAD_REQUEST);
    }
    return this.memorialService.createMemorialFlag(
      userId,
      request.type,
      request.referenceId,
      request.reason,
    );
  }

  @Get('published/:slug/flag')
  async getMemorialFlags(
    @Session() session: SessionContainer,
    @Param('slug') slug: string,
  ): Promise<MemorialFlag[]> {
    const userId = session.getUserId();
    return this.memorialService.getMemorialFlags(userId, slug);
  }

  @Patch('flag/:flagId')
  async handleMemorialFlag(
    @Session() session: SessionContainer,
    @Param('flagId') flagId: string,
    @Body() { status }: { status: MemorialFlagStatus.APPROVED | MemorialFlagStatus.REJECTED },
  ): Promise<void> {
    const userId = session.getUserId();
    return this.memorialService.handleMemorialUpdateFlag(userId, flagId, status);
  }

  @Get('ai-greeting/:memorialId')
  async getMemorialAiGreeting(
    @Param('memorialId') memorialId: string,
    @Session() session: SessionContainer,
  ) {
    await this.memorialService.verifyMemorialOwnership(session.getUserId(), memorialId);
    return this.aiService.getAiGreeting(memorialId);
  }

  @Delete('ai-greeting/:memorialId')
  async deleteMemorialAiVoice(
    @Param('memorialId') memorialId: string,
    @Session() session: SessionContainer,
  ): Promise<void> {
    await this.memorialService.verifyMemorialOwnership(session.getUserId(), memorialId);
    return this.aiService.resetAiGreetingCreation(memorialId);
  }

  @Post('ai-greeting/voice/:memorialId')
  @UseInterceptors(FilesInterceptor('files'))
  async createMemorialVoice(
    @Param('memorialId') memorialId: string,
    @Session() session: SessionContainer,
    @UploadedFiles() files: Express.Multer.File[],
    @Query('locale') locale?: 'tr' | 'en',
  ): Promise<{ path: string }> {
    await this.memorialService.verifyMemorialOwnership(session.getUserId(), memorialId);
    const voiceId = await this.aiService.createMemorialVoice(memorialId, files);
    const localeToUse = locale && ['tr', 'en'].includes(locale) ? locale : 'en';
    const path = await this.aiService.createMemorialGreetingAudio(memorialId, voiceId, localeToUse);
    return { path: generateAssetUrl(path) };
  }

  @Post('ai-greeting/image/:memorialId')
  @UseInterceptors(
    FileInterceptor('image', {
      limits: {
        files: 1,
      },
    }),
  )
  async uploadAiGreetingImage(
    @Param('memorialId') memorialId: string,
    @Session() session: SessionContainer,
    @UploadedFile() image: Express.Multer.File,
  ): Promise<{ path: string }> {
    await this.memorialService.verifyMemorialOwnership(session.getUserId(), memorialId);
    const path = await this.aiService.uploadAiGreetingImage(memorialId, image);
    return { path: generateAssetUrl(path) };
  }

  @PublicAccess()
  @Get('published/:slug/top-contributors/candles')
  async getTopCandleContributors(@Param('slug') slug: string): Promise<TopContributor[]> {
    return this.memorialService.getTopCandleContributors(slug);
  }

  @PublicAccess()
  @Get('published/:slug/top-contributors/donations')
  async getTopDonors(@Param('slug') slug: string): Promise<TopContributor[]> {
    return this.memorialService.getTopDonors(slug);
  }

  @PublicAccess()
  @Get('published/:slug/top-contributors/flowers')
  async getTopFlowerContributors(@Param('slug') slug: string): Promise<TopContributor[]> {
    return this.memorialService.getTopFlowerContributors(slug);
  }

  @PublicAccess()
  @Get('published/:slug/top-contributors/trees')
  async getTopTreePlanters(@Param('slug') slug: string): Promise<TopContributor[]> {
    return this.memorialService.getTopTreePlanters(slug);
  }
}
